state/executor.go

```diff
@@ -593,21 +603,42 @@ func (t *Transition) Call2(
 }
 
 func (t *Transition) run(contract *runtime.Contract, host runtime.Host) *runtime.ExecutionResult {
-	// check allow list (if any)
-	if t.deploymentAllowlist != nil && t.deploymentAllowlist.Addr() == contract.CodeAddress {
-		return t.deploymentAllowlist.Run(contract, host, &t.config)
+	// check contract deployment allow list (if any)
+	if t.deploymentAllowList != nil && t.deploymentAllowList.Addr() == contract.CodeAddress {
+		return t.deploymentAllowList.Run(contract, host, &t.config)
+	}
+
+	// check contract deployment block list (if any)
+	if t.deploymentBlockList != nil && t.deploymentBlockList.Addr() == contract.CodeAddress {
+		return t.deploymentBlockList.Run(contract, host, &t.config)
 	}
 
+	// check txns access lists, allow list takes precedence over block list
 	if t.txnAllowList != nil {
 		if t.txnAllowList.Addr() == contract.CodeAddress {
 			return t.txnAllowList.Run(contract, host, &t.config)
 		}
 
 		if contract.Caller != contracts.SystemCaller {
-			txnAllowListRole := t.txnAllowList.GetRole(contract.Caller)
-			if !txnAllowListRole.Enabled() {
+			role := t.txnAllowList.GetRole(contract.Caller)
+			if !role.Enabled() {
 				return &runtime.ExecutionResult{
-					Err: fmt.Errorf("caller not authorized"),
+					GasLeft: 0,
+					Err:     runtime.ErrNotAuth,
+				}
+			}
+		}
+	} else if t.txnBlockList != nil {
+		if t.txnBlockList.Addr() == contract.CodeAddress {
+			return t.txnBlockList.Run(contract, host, &t.config)
+		}
+
+		if contract.Caller != contracts.SystemCaller {
+			role := t.txnBlockList.GetRole(contract.Caller)
+			if role == addresslist.EnabledRole {
```

Comment:
Should we check `AdminRole` here as well (it takes precedence over `EnabledRole` to my understanding)?

```suggestion
			if role.Enabled() {
```

Reply:
With your suggestion we'll block admins or the ones in the list, we don't want that, `AdminRole` will simply pass the check forward, that's what we want.

Reply:
I suggested this since the same way it is checked for allowed list:
if an account has either `AdminRole` or `EnabledRole` it is considered to be whitelisted. Although not sure if that makes sense either. To me `AdminRole` is allowed to set roles, but probably should not be implicitly considered as enabled. Not sure, what are your thoughts here? :thinking:

Reply:
I think the expectation is that Admin is enabled by default. That said, I think semantically `Enabled` doesn't makes sense in BlockLists.

Do you feel it's worth to add also `DisabledRole` and use that for BlockLists?

Reply:
No need to add another role, it is ok like this: if an account is enabled in block list => it is disabled.